### PR TITLE
MAIN - adds medical sticker encryption charlie

### PR DESCRIPTION
**Pull request by paschalidi**
Heya, @lanwen and @Weetbix. 

Hope you enjoy the weekend that is coming! 
This will add the new version of the encryption for the medical id

### CODE REVIEW

File: src/lib/MedStickerEncryptionV2.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+const gcm = create("AES-GCM");
+
+/**
+ *
+ * @param secret
+ * @param salt
+ * @param options
+ * @returns {PromiseLike<CryptoKeyPair> | PromiseLike<CryptoKey> | PromiseLike<CryptoKeyPair | CryptoKey>}
+ */
+export async function hash(secret, salt, options) {
```
Comment (lanwen) [inline src/lib/MedStickerEncryptionV2.js]: please hide options from the api - they should be part of the lib

File: src/lib/MedStickerEncryptionV2.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ *
+ * @param array
+ * @returns {{cryptoKey: ArrayBufferLike, accessKey: ArrayBufferLike}}
+ */
+export function splitKeys(array) {
+  const arrayLength = array.length;
+  const cryptoKey = array.slice(0, arrayLength / 2);
+  const accessKey = array.slice(arrayLength / 2, arrayLength);
```
Comment (lanwen) [inline src/lib/MedStickerEncryptionV2.js]: lets name it as in rfc key and fingerprintFile

**Comment by lanwen** [comment]
please add all the consts used into the lib

File: src/lib/MedStickerEncryptionV2.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+/**
+ *
+ * @param hashed
+ * @returns {Promise<string>}
+ */
+export function fingerprintSecret(hashed) {
+  const base64EncodedFingerprintSecret = encode(arrayBufferToString(hashed));
+  return `${CHARLIE}-sha256:${base64EncodedFingerprintSecret}`;
```
Comment (lanwen) [inline src/lib/MedStickerEncryptionV2.js]:
```suggestion
  return `${CHARLIE}:${base64EncodedFingerprintSecret}`;
```
not sha256 anymore

Comment (lanwen) [reply src/lib/MedStickerEncryptionV2.js]: please align with client is that base64 or hex

File: src/lib/MedStickerEncryptionV2.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    key: new Uint8Array(cryptoKey).buffer,
+    fingerprintFile: new Uint8Array(fingerprintFile).buffer
+  };
+}
+
+/**
+ *
+ * @returns {string}
+ */
+export function generateRandomAesIv() {
```
Comment (lanwen) [inline src/lib/MedStickerEncryptionV2.js]: is that really needed here? I think it could just throw unimplemented error - since you don't need that even in tests

File: src/lib/MedStickerEncryptionV2.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+ * @returns {{cryptoKey: ArrayBufferLike, accessKey: ArrayBufferLike}}
+ */
+export function splitKeys(array) {
+  const arrayLength = array.length;
+  const cryptoKey = array.slice(0, arrayLength / 2);
+  const fingerprintFile = array.slice(arrayLength / 2, arrayLength);
+
+  return {
+    key: new Uint8Array(cryptoKey).buffer,
+    fingerprintFile: new Uint8Array(fingerprintFile).buffer
```
Comment (lanwen) [inline src/lib/MedStickerEncryptionV2.js]: should be a string same way as other fingerprint

Comment (paschalidi) [reply src/lib/MedStickerEncryptionV2.js]: ah cool

File: src/lib/MedStickerEncryptionV2.js (hunk 4 above)
Comment (paschalidi) [reply src/lib/MedStickerEncryptionV2.js]: > I think it could just throw unimplemented error - since you don't need that even in tests

this will generate the `initializationVector` with some random values and its used in the tests to generate `iv`s so that the `encrypt` function can be fed with it. 
Examples of how its being used in the tests 
[here](https://github.com/VivyTeam/krypt-web/pull/73/files/82ff7b337f23b009024436d4fcbdd67d48388a1e#diff-8c6a8dc4a932988d10415763da2a7dadR36)
and [here](https://github.com/VivyTeam/krypt-web/pull/73/files/82ff7b337f23b009024436d4fcbdd67d48388a1e#diff-8c6a8dc4a932988d10415763da2a7dadR23). 

Probably you mean something i dont get?

File: src/lib/MedStickerEncryptionV2.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
+import create from "./factory";
```
Comment (Weetbix) [inline src/lib/MedStickerEncryptionV2.js]: Now we have 3 Med Sticker versions,

Version 1 and 2 named A and B implemented in `MedStickerEncrypted.js`

And Version 3 named C and implemented in `MedStickerEncryptedV2.js` 

This is really confusing

Comment (Weetbix) [reply src/lib/MedStickerEncryptionV2.js]: Why arent we doing this through the same API as the first 2 versions?

Comment (paschalidi) [reply src/lib/MedStickerEncryptionV2.js]: @Weetbix you mean that these all should be living in the same directory or?
I agree that naming can be better, any suggestions?

File: src/lib/MedStickerEncryptionV2.js (hunk 4 above)
Comment (lanwen) [reply src/lib/MedStickerEncryptionV2.js]: you don't need random values in tests - just use some predefined bytes.

Thus you can skip this method and limit the scope of the lib

Comment (paschalidi) [reply src/lib/MedStickerEncryptionV2.js]: yeah but in the real world i would need that function, why shouldnt we have it in one place for all our projects that might be using the lib?

Therefore when we want to update the function with for instance a different implementation we just do it in the lib and not in all our projects?

File: src/lib/MedStickerEncryptionV2.js (hunk 6 above)
Comment (lanwen) [reply src/lib/MedStickerEncryptionV2.js]: The suggestion was to have a separate charlie class

File: src/lib/MedStickerEncryptionV2.js (hunk 4 above)
Comment (paschalidi) [reply src/lib/MedStickerEncryptionV2.js]: I can ofc delete it no big deal but I think we might wanna have it here.

File: src/lib/MedStickerEncryptionV2.js (hunk 6 above)
Comment (lanwen) [reply src/lib/MedStickerEncryptionV2.js]: @Weetbix its not fully compatible with the old api even on signatures level, not speaking about implementation

Comment (Weetbix) [reply src/lib/MedStickerEncryptionV2.js]: v3 is living in a v2 file... would be better if instead you stick to the naming (and keep the version correct) like `medStickerCharlie.js` 

I dont mind if the implementations are in separate files, but currently you would have to import decrypt from 2 different places and know what version you want. 

Why arent we updating this: https://github.com/VivyTeam/krypt-web/blob/master/src/lib/MedStickerEncryption.js#L23

File: src/lib/MedStickerEncryptionV2.js (hunk 4 above)
Comment (lanwen) [reply src/lib/MedStickerEncryptionV2.js]: until we gonna activate the sticker in the web (don't think that would happen anytime) - we won't need that method. If we don't need that - why to keep it?

Comment (paschalidi) [reply src/lib/MedStickerEncryptionV2.js]: ah yeah okay didnt think of that this way. you are right. 

One last thing, there was the idea that the libraries have exactly the same functions exported and the same api for someone to use it, is it now given that, fine, if the other two libs have this function and web doesnt?

Comment (lanwen) [reply src/lib/MedStickerEncryptionV2.js]: so you actually already have a different api :) but my idea was to keep the method - just don't implement that, so that signatures will be the same with only difference of the internal details

File: src/lib/MedStickerEncryptionV2.js (hunk 6 above)
Comment (lanwen) [reply src/lib/MedStickerEncryptionV2.js]: @Weetbix iv now comes from the server, besides the qr code content secret from the backend is now required. I can introduce you new security concept to give you more context

File: src/lib/MedStickerEncryptionV2.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+import { CHARLIE } from "./constants";
+
+const scrypt = create("scrypt");
+const gcm = create("AES-GCM");
+
+/**
+ *
+ * @param secret
+ * @param salt
+ * @returns {PromiseLike<CryptoKeyPair> | PromiseLike<CryptoKey> | PromiseLike<CryptoKeyPair | CryptoKey>}
```
Comment (Weetbix) [inline src/lib/MedStickerEncryptionV2.js]: Looking at the implementation, this return type is way off what it really will be. Can you check what it really is and use the correct type here. 

Infact, the implementation of the scrypt call in our factory is incorrect:
https://github.com/VivyTeam/krypt-web/blob/master/src/lib/factory.js#L212
https://github.com/dchest/scrypt-async-js#scryptpassword-salt-options-callback

It's only working right now because no one has passed a non-zero value of `interruptStep` to it. 
In the factory its treated as always being a sync call. As soon as someone passes an interrupt step its going to return null.

File: src/lib/MedStickerEncryptionV2.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
+import create from "./factory";
+import { arrayBufferToString, encode } from "./utilities";
+import { CHARLIE } from "./constants";
+
+const scrypt = create("scrypt");
+const gcm = create("AES-GCM");
+
+/**
+ *
```
Comment (Weetbix) [inline src/lib/MedStickerEncryptionV2.js]: Sometimes functions dont need comments, sometimes they do. But they never need blank lines where the comment should be, so please decide which ones should be described and not. For the ones that not, remove the blank lines

File: src/lib/MedStickerEncryptionV2.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+ * @param secret
```
Comment (Weetbix) [inline src/lib/MedStickerEncryptionV2.js]: For all these params please either add a description, or their type, or remove them. Right now they are not useful at all.

File: src/lib/MedStickerEncryptionV2.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    return await scrypt.generateKey(secret, salt);
+  } catch (e) {
+    throw new Error("HashFailed");
+  }
+}
+
+/**
+ *
+ * @param hashed
+ * @returns {Promise<string>}
```
Comment (Weetbix) [inline src/lib/MedStickerEncryptionV2.js]: return type is incorrect. Please check the return types on all the functions to ensure they are right

File: src/lib/MedStickerEncryptionV2.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+/**
+ *
+ * @param toEncryptBytes
+ * @param key
+ * @param iv
+ * @returns {PromiseLike<ArrayBuffer>}
+ */
+export async function encrypt(toEncryptBytes, key, iv) {
```
Comment (Weetbix) [inline src/lib/MedStickerEncryptionV2.js]: In all the other web krypt apis we take the data/bytes last. Is there a reason they moved to the front for this? i would rather keep them all consistent

File: src/lib/MedStickerEncryptionV2.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+/**
+ *
+ * @param encryptedData
+ * @param key
+ * @param iv
+ * @returns {PromiseLike<ArrayBuffer>}
+ */
+export async function decrypt(encryptedData, key, iv) {
```
Comment (Weetbix) [inline src/lib/MedStickerEncryptionV2.js]: the above function param is `toEncryptBytes` and this one is `encryptedData`. Can't we make it `encryptedBytes`? they should be consistent

File: src/lib/constants.js [hunk 13]
```diff
@@ -1,2 +1,4 @@
 export const ADAM = "adam";
 export const BRITNEY = "britney";
+export const CHARLIE = "charlie";
+export const charlieStaticSalt = "5f1288159017d636c13c1c1b2835b8a871780bc2";
```
Comment (Weetbix) [inline src/lib/constants.js]: Please use the constants naming convention seen above: `CHARLIE_STATIC_SALT`

File: src/lib/MedStickerEncryptionV2.js (hunk 6 above)
Comment (paschalidi) [reply src/lib/MedStickerEncryptionV2.js]: Have we concluded that this file should be renamed MedStickerEncryptionV1.js?

File: src/lib/MedStickerEncryptionV2.js (hunk 7 above)
Comment (paschalidi) [reply src/lib/MedStickerEncryptionV2.js]: do you recommend that in the factory this should be implemented as an `async`?

File: src/lib/MedStickerEncryptionV2.js (hunk 6 above)
Comment (Weetbix) [reply src/lib/MedStickerEncryptionV2.js]: This should be named `medStickerEncryptionCharlie` 

I would like to know how you plan to expose this API publically too, what the interface will look like? as in this PR its not usable, as its not exported in https://github.com/VivyTeam/krypt-web/blob/master/src/index.js

File: src/lib/MedStickerEncryptionV2.js (hunk 7 above)
Comment (Weetbix) [reply src/lib/MedStickerEncryptionV2.js]: Either generateKey needs to return a promise, or you need to override the option to always be 0 no matter what the user inputs. 

Look at the signature of the other generateKeys to decide the best option

**Comment by paschalidi** [comment]
@Weetbix @lanwen I took a close look into all your comments, hoped didnt miss anything. 
Please have a look and let me know your thoughts! 

As always thanks for the very onto the point feedback. 👍

File: src/lib/MedStickerEncryptionCharlie.js [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * Given a hashed string, returns a concatenated string
+ * that contains the version of the encryption its being
+ * used and the incoming string in hex format
+ * @param hashed {string}
+ * @returns {string}
+ */
+
+export function fingerprintSecret(hashed) {
```
Comment (lanwen) [inline src/lib/MedStickerEncryptionCharlie.js]: its not only secret fingerprint

File: src/tests/MedStickerEncryptionV2.test.js [hunk 15]
```diff
@@ -0,0 +1,66 @@
+import {
```
Comment (lanwen) [inline src/tests/MedStickerEncryptionV2.test.js]: please rename the test file accordingly

File: src/lib/MedStickerEncryptionV2.js (hunk 7 above)
Comment (paschalidi) [reply src/lib/MedStickerEncryptionV2.js]: can we please check this together on what the returned value type should be @Weetbix ?

**Comment by paschalidi** [comment]
@Weetbix @lanwen sorry realised I missed some docs so I got it updated.

File: src/lib/utilities.js [hunk 16]
```diff
@@ -63,3 +63,30 @@ export const concatenateUint8Arrays = (...arrays) => {
...
+export const arrayBufferToHex = buffer => {
+  if (buffer.buffer instanceof ArrayBuffer && buffer.byteLength !== undefined) {
+    throw new TypeError("Expected input to be an ArrayBuffer");
+  }
+  return Array.prototype.map
+    .call(new Uint8Array(buffer), x => `00${x.toString(16)}`.slice(-2))
+    .join("");
+};
+
+export const hexToArrayBuffer = hex => {
```
Comment (lanwen) [inline src/lib/utilities.js]: where do you use this function?

Comment (paschalidi) [reply src/lib/utilities.js]: I wrote a simple test to validate that the function `arrayBufferToHex` works as expected

**Comment by lanwen** [comment]
can you add resulting constants into tests?

**Comment by paschalidi** [comment]
what is "resulting constants"?

File: src/tests/MedStickerEncryptionCharlie.test.js [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+    const cryptoKey = await gcm.importKey(key);
+
+    const encryptedArrayBuffer = await encrypt(cryptoKey, iv, buffer);
+    const arrayBufferData = await decrypt(cryptoKey, iv, encryptedArrayBuffer);
+    const result = arrayBufferToString(arrayBufferData);
+
+    expect(result).toEqual(originalString);
+  });
+
+  it("should throw error on decrypt when iv is different that on used for encryption", async () => {
```
Comment (Weetbix) [inline src/tests/MedStickerEncryptionCharlie.test.js]:
```suggestion
  it("should throw error on decrypt when the IV used is not the same as the one used for encryption", async () => {

```

File: src/tests/MedStickerEncryptionCharlie.test.js [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    try {
+      await decrypt(cryptoKey, anotherIv, encryptedArrayBuffer);
+    } catch (err) {
+      error = err;
+    } finally {
+      expect(error.message).toEqual("DecryptionFailed");
+    }
+  });
+
+  it("should return a string in the form of charlie:{fingerprint}", async () => {
```
Comment (Weetbix) [inline src/tests/MedStickerEncryptionCharlie.test.js]: these are marked as async but you are not awaiting anything.. (also in the next function)

File: src/lib/utilities.js (hunk 16 above)
Comment (Weetbix) [inline src/lib/utilities.js]: This is only used in the test, i recommend we just get rid of this (see comment in test file)

File: src/tests/utilities.test.js [hunk 19]
```diff
@@ -16,4 +21,12 @@ describe("basic utilities", () => {
 
     expect(result).toEqual(originalArrayBuffer);
   });
+
+  it("should convert an ArrayBuffer to Hex and back. Result should be the equal", async () => {
```
Comment (Weetbix) [inline src/tests/utilities.test.js]: This test is also testing your test-only util function and relying on the implementation of that. 

Why don't we just totally simply this test and write:
```
const original = new Uint8Array([ 0, 1, 10, 16, 240, 255 ]);
const expected = "00010a10f0ff";
expect(arrayBuffereToHex(original.buffer)).toEqual(expected);
```

File: src/tests/MedStickerEncryptionCharlie.test.js (hunk 17 above)
Comment (Weetbix) [reply src/tests/MedStickerEncryptionCharlie.test.js]: oh dont commit this directly, seems there is an extra new line

**Comment by lanwen** [comment]
like hash and fingerprint something, get the result and check that we actually produce that string later in tests, using what other platform libs providing

File: src/lib/utilities.js (hunk 16 above)
Comment (lanwen) [reply src/lib/utilities.js]: then it shouldn't be part of the lib, only part of the tests

File: src/tests/MedStickerEncryptionCharlie.test.js [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+  splitKeys,
+  fingerprint
+} from "../lib/MedStickerEncryptionCharlie";
+import { arrayBufferToString, stringToArrayBuffer } from "../lib/utilities";
+import create from "../lib/factory";
+import { CHARLIE_STATIC_SALT } from "../lib/constants";
+
+const gcm = create("AES-GCM");
+
+describe("second version of MedStickerEncryption", () => {
```
[the remark this answers is not in the document]
Comment (lanwen) [reply src/tests/MedStickerEncryptionCharlie.test.js]: actually the first public, so better to stick with charlie

File: src/tests/MedStickerEncryptionCharlie.test.js (hunk 18 above)
Comment (paschalidi) [reply src/tests/MedStickerEncryptionCharlie.test.js]: now that we changed the `hash` to be async this will not be the case anymore.

File: src/lib/utilities.js (hunk 16 above)
Comment (paschalidi) [reply src/lib/utilities.js]: I need this func to use it also in the contract tests. 

For instance to validate that the lib works between the different sources (ios and android) what we will do is that they will encrypt some sort of data, then will give me the result o this encryption as a `hex` value since passing around ArrayBuffers is not working well for us. 

Then I will have this hex value and try to decrypt it to see if my decryption works as expected. In that case i will have to convert the hex to ArrayBuffer and therefore I will need something like this. 

What you think?

File: src/tests/utilities.test.js (hunk 19 above)
Comment (paschalidi) [reply src/tests/utilities.test.js]: please look here https://github.com/VivyTeam/krypt-web/pull/73#discussion_r297198204, I might be wrong, in this case please put me right!

File: src/index.js [hunk 21]
```diff
@@ -6,12 +6,20 @@ export {
...
 } from "./lib/EHREncryption";
+
 export {
   decrypt as medDecrypt,
   accessSignature as medAccessSignature,
   deriveKey as medDeriveKey
 } from "./lib/MedStickerEncryption";
 
+export {
+  decrypt as charlieDecrypt,
```
[the remark this answers is not in the document]
Comment (paschalidi) [reply src/index.js]: how do `medStickerFingerprint`, `medStickerSplitKeys`, `medStickerHash`, `medStickerDecrypt` sound?

File: src/lib/utilities.js (hunk 16 above)
Comment (lanwen) [reply src/lib/utilities.js]: still should be part of the tests and actually inlined as @Weetbix recommended

Comment (paschalidi) [reply src/lib/utilities.js]: so we will have this private function inside the tests to transform from hex to array buffer? 
but don't we want to test that this function actually does what it suppose to and therefore get tested? 
Adding a function to the tests that will implement some logic can make our test pass or fail unpredictably. For instance it can be that this private function has a bug and therefore the tests are passing when it should fail. 
I think it makes sense that we test this function before we use it in the tests. 

What you think?

**Comment by paschalidi** [comment]
As we talked privately with both of you will have the contract test in a different PR once this is closed.

File: src/lib/utilities.js (hunk 16 above)
Comment (lanwen) [reply src/lib/utilities.js]: we just ask you to get rid of it and inline the result where needed.

In case you want to have something for tests - move it to tests scope, and create test nearby, it shouldn't be publicly available.

**Comment by Weetbix** [comment]
@lanwen @paschalidi 



Hi peeps.

I have a proposal for the public API of the library which will allow us to avoid
these weird naming of versios in past and future versions. 

Current public API in this PR:

```javascript
// ADAM + BRITNEY
medDecrypt({ key, iv, version }, data)) : Promise<ArrayBuffer>
medAccessSignature({ key, iv, version }, salt)
medDeriveKey(code, pin, version) : Promise<{ key, iv, version }>


// CHARLIE
charlieDecrypt(key, iv, encryptedBytes)

// for key derivation...
charlieHash(secret, salt)
charlieSplitKeys(array) : {fingerprintFile, key}
charlieFingerprint(ArrayBuffer): string
```

Proposed public API 
```javascript
// Each version has associated options object
medDecrypt(VERSION, options, data) : Promise<ArrayBuffer>
medEncrypt(VERSION, options ) : Promise<ArrayBuffer>

// options for different versions:
ADAM : { key, iv }
BRITNEY : { key, iv }
CHARLIE : { key, iv }

// key derivation functions, where the options really do
// change between versions 
medDeriveKey(VERSION, options) : Promise<{ key, iv? }>

// options for different versions
ADAM : { code, pin }
BRITNEY : { code, pin }
CHARLIE : { pin, pinSecret }

// example implementation of derive key. we can do the samee for medDecrypt
import { adamDeriveKey } from "./versions/adam";
import { charlieDeriveKey } from "./versions/charlie";
function medDeriveKey(version, options)
{
  switch(version)
  {
    case ADAM: 
      const { code, pin } = options;
      return adamDeriveKey(code, pin, ADAM)
    //blah blah
    case CHARLIE:
      const { pin, pinSecret } = options;
      // hash and salt look like they can be done inside the lib
      return charlieDeriveKey(pin, pinSecret)
  }
}


// Then, example decrypts:
import { medDecrypt, medDeriveKey } from "@vivy/krypt-web"
import { ADAM, BRITNEY, CHARLIE } from "@vivy/krypt-web/versions";

// adam
const { key, iv } = await medDeriveKey( ADAM, { code : 'blah', pin = 'blah' } )
const data = await medDecrypt( ADAM, { key, iv }, data );

// britney
const { key, iv } = await medDeriveKey( BRITNEYE, { code : 'blah', pin = 'blah' } )
const data = await medDecrypt( BRITNEY, { key, iv }, data );

// Charlie
const { key } = await medDeriveKey( CHARLIE, { pin : 'blah', pinSecret = 'blah' } )
// IV in your tests is empty... not sure where its meant ot come fropm
const iv = new ArrayBuffer(128);
const data = await medDecrypt( CHARLIE, { key, iv }, data );
```

In summary we only expose 1 encrypt/decrypt/derive key function which is then given
a different veresion and options object based on the versions. This way we are keeping
a consistent encrypt/decrypt/derive key entry point that can change with different
versions and options. 

Alternatively we could change them all individual exports as the consumer needs
to know the version they are targetting anyway: 

```javascript
import { decrypt, deriveKey } from "@vivy/krypt-web/medSticker/version/adam";

const { key } = await deriveKey({ pin : 'blah', pinSecret = 'blah' })
const iv = new ArrayBuffer(128);
const data = await decrypt({ key, iv }, data);
```

However the advantage of the first method is that changes will be minimal between versions with identical APIs (ie adam/britney)

thoughts?

**Comment by lanwen** [comment]
@Weetbix the issue with that is adam and britney are 1 step flow and charlie is 2 step flow,

but generally I'd like to replace one constant

**Comment by Weetbix** [comment]
@lanwen ah... but cant we still expose `deriveKey` and `decrypt` behind this API. decrypt for sure, because it always expects an key and iv to be passed to it. 

Derive key kind of because its only 1 part for all encryptions so far, but it also happens to return the IV for adam and britney. 

Maybe then we should split it into 3: 
```
deriveKey(VERSION, { ...options }) 
deriveIV(VERSION, {...options} )
decrypte(VERSION, { key, iv } )
```

File: src/lib/MedStickerEncryptionCharlie.js [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+ * from the second bit creates a string that
+ * contains a secret but also indicates the version of the algorithm used
+ * @param array {Array}
+ * @returns {{fingerprintFile: string, key: ArrayBufferLike}}
+ */
+export function splitKeys(array) {
+  const arrayLength = array.length;
+  const key = array.slice(0, arrayLength / 2);
+  const fingerprintFile = array.slice(arrayLength / 2, arrayLength);
+  const fingerprintArrayBuffer = new Uint8Array(fingerprintFile).buffer;
```
Comment (lanwen) [inline src/lib/MedStickerEncryptionCharlie.js]: please inline this variable, 

btw isn't that already array buffer after splitting an array buffer?

File: src/lib/MedStickerEncryptionCharlie.js [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * Takes an array, splits it in half,
+ * from the first bit creates an ArrayBuffer
+ * from the second bit creates a string that
+ * contains a secret but also indicates the version of the algorithm used
+ * @param array {Array}
+ * @returns {{fingerprintFile: string, key: ArrayBufferLike}}
+ */
+export function splitKeys(array) {
+  const arrayLength = array.length;
```
Comment (lanwen) [inline src/lib/MedStickerEncryptionCharlie.js]: don't think this is needed at all - `array.length` is just 1 symbol longer than `arrayLength`, but you've completely mitigated that drawback with `const` - since you use `arrayLength` 3 times only :) In fact this adds 2 symbols more to your code, than just not using it

File: src/lib/MedStickerEncryptionCharlie.js (hunk 22 above)
Comment (paschalidi) [reply src/lib/MedStickerEncryptionCharlie.js]: the incoming value is array. on my next push this will be clearer.

**Comment by paschalidi** [comment]
@Weetbix @lanwen I very carefully saw your comments and made sure all of it is addressed, please be so kind and let me know if I missed anything. 


#### Some minor comments that might be helpful
- I checked and triple checked the types :sweat_smile:, hopefully this is now correct.
- `ArrayBufferLike`:  when you have `new Uint8Array(key).buffer;` my editor would name that `ArrayBufferLike`. Lets discuss if this is how we wanna implement it. :white_check_mark:
- We did sit down with John and the conclusion we reached, was that ` scrypt.generateKey` should also return a promise so that the `api` is consistent. This is why now the function `deriveKey` has become an `async` one.

- The correct options for the `script.generateKey` of this version is `{ dkLen: 64, r: 10 }` I had this wrong before so now it is fixed.

- Also the `hash` function now returns an `arrayBuffer`. Before it  used to return `Array` since this is what `scrypt.generateKey` returns. I find this a bit confusing since we always manipulate `ArrayBuffer` here. One can argue that `scrypt.generateKey` should return itself an `ArrayBuffer` . Lets discuss if this is how we wanna implement it. :white_check_mark:

File: src/lib/MedStickerEncryptionCharlie.js [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+ * Takes an array, splits it in half,
+ * from the first bit creates an ArrayBuffer
+ * from the second bit creates a string that
+ * contains a secret but also indicates the version of the algorithm used
+ * @param arrayBuffer {ArrayBufferLike | ArrayBuffer}
+ * @returns {{fingerprintFile: string, key: ArrayBufferLike}}
+ */
+export function splitKeys(arrayBuffer) {
+  const key = arrayBuffer.slice(0, arrayBuffer.byteLength / 2);
+  const fingerprintFile = arrayBuffer.slice(
```
Comment (lanwen) [inline src/lib/MedStickerEncryptionCharlie.js]:
```suggestion
  const fingerprint = fingerprint(arrayBuffer.slice(
```

and later just 
`fingerprintFile: fingerprint`

otherwise too many fingerprints around

File: src/lib/factory.js [hunk 25]
```diff
@@ -205,7 +205,7 @@ export default algorithm => {
          * @param password {string}
          * @param salt {string}
          * @param options {object}
-         * @returns {arrayBuffer}
+         * @returns {Array}
```
Comment (paschalidi) [inline src/lib/factory.js]: may I correct this? It was never an ArrayBuffer

File: src/lib/utilities.js [hunk 26]
```diff
@@ -63,3 +63,15 @@ export const concatenateUint8Arrays = (...arrays) => {
...
 };
+
+export const arrayBufferToHex = buffer => {
+  if (buffer.buffer instanceof ArrayBuffer && buffer.byteLength !== undefined) {
+    throw new TypeError("Expected input to be an ArrayBuffer");
+  }
+  const hex = new Uint8Array(buffer);
+  return hex.reduce(
+    (accumulatedHexString, current) =>
+      accumulatedHexString + current.toString(15).padStart(2, "0"),
```
Comment (Weetbix) [inline src/lib/utilities.js]: This radix is wrong it should be 16. Subsequently this function is broken. Try converting 255 to FF. 

Please add a failing test case with edge cases and then fix this

File: src/tests/utilities.test.js [hunk 27]
```diff
@@ -16,4 +20,11 @@ describe("basic utilities", () => {
...
+    const original = new Uint8Array([0, 1, 10, 16, 1, 123, 33, 14, 15]);
```
Comment (Weetbix) [inline src/tests/utilities.test.js]: Im curious why you changed test data around, in my example we are testing the boundaries and it would have caught the incorrect radix used in the toHex function above

File: src/index.js [hunk 28]
```diff
@@ -6,6 +6,7 @@ export {
   decrypt as ehrDecrypt,
   encrypt as ehrEncrypt
 } from "./lib/EHREncryption";
+
```
[the remark this answers is not in the document]
Comment (paschalidi) [reply src/index.js]: well this is for uniformity. all the other exports have a space in between. still needs to be removed you think?>

File: src/lib/MedStickerEncryptionCharlie.js [hunk 29]
```diff
@@ -0,0 +1,80 @@
...
+  };
+}
+
+/**
+ * @param key {CryptoKey}
+ * @param iv {Uint8Array}
+ * @param toEncryptBytes {ArrayBuffer}
+ * @returns {Promise<ArrayBuffer>}
+ */
+export async function encrypt(key, iv, toEncryptBytes) {
```
[the remark this answers is not in the document]
Comment (paschalidi) [reply src/lib/MedStickerEncryptionCharlie.js]: but i made this change. now decrypt takes `encryptedBytes`  and this func takes `toEncryptBytes`. What is not addressed?

[the remark this answers is not in the document]
Comment (Weetbix) [reply src/lib/MedStickerEncryptionCharlie.js]: Ah yes! Sorry for the confusion

File: src/index.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (Weetbix) [reply src/index.js]: yes please remove as its the only change in the file, and can lead to confusion when reading the commit history later as you'll read "Adds medical sticker encryption charlie" as having changed the index.js in this PR where we will not change it. This is fine obviously if you are adding other changes in the same file, but we have generally never formatted random other files in our PRs

Comment (paschalidi) [inline src/index.js]:
```suggestion
```

[the remark this answers is not in the document]
Comment (paschalidi) [reply src/index.js]: thanks for explaining John. Will follow that from now on

File: src/lib/MedStickerEncryptionCharlie.js (hunk 29 above)
[the remark this answers is not in the document]
Comment (paschalidi) [reply src/lib/MedStickerEncryptionCharlie.js]: No probs. I think that PR is a bit crazy. Sorry if I miss some things @Weetbix on my behalf as well. I try to make it as smooth as possible

File: src/lib/MedStickerEncryptionV2.js (hunk 5 above)
Comment (paschalidi) [reply src/lib/MedStickerEncryptionV2.js]: afaic this has been addressed. please resolve or correct me if i am worng

File: src/lib/MedStickerEncryptionCharlie.js (hunk 14 above)
Comment (paschalidi) [reply src/lib/MedStickerEncryptionCharlie.js]: afaic this has been addressed. please resolve or correct me if i am worng
